diff --git a/src/AV/Output/AudioEncoder.cpp b/src/AV/Output/AudioEncoder.cpp
index 70752b3..2bbc07a 100644
--- a/src/AV/Output/AudioEncoder.cpp
+++ b/src/AV/Output/AudioEncoder.cpp
@@ -34,7 +34,7 @@ const std::vector<AudioEncoder::SampleFormatData> AudioEncoder::SUPPORTED_SAMPLE
 
 const unsigned int AudioEncoder::DEFAULT_FRAME_SAMPLES = 1024;
 
-AudioEncoder::AudioEncoder(Muxer* muxer, AVStream* stream, AVCodecContext *codec_context, AVCodec* codec, AVDictionary** options)
+AudioEncoder::AudioEncoder(Muxer* muxer, AVStream* stream, AVCodecContext *codec_context, const AVCodec* codec, AVDictionary** options)
 	: BaseEncoder(muxer, stream, codec_context, codec, options) {
 
 #if !SSR_USE_AVCODEC_ENCODE_AUDIO2
@@ -53,6 +53,9 @@ AudioEncoder::AudioEncoder(Muxer* muxer, AVStream* stream, AVCodecContext *codec
 }
 
 AudioEncoder::~AudioEncoder() {
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+	av_channel_layout_uninit(&GetCodecContext()->ch_layout);
+#endif
 	StopThread();
 }
 
@@ -70,9 +73,6 @@ AVSampleFormat AudioEncoder::GetSampleFormat() {
 
 unsigned int AudioEncoder::GetChannels() {
 #if LIBAVCODEC_VERSION_MAJOR < 61
-#if SSR_USE_AV_CHANNEL_LAYOUT
-	return GetCodecContext()->ch_layout.nb_channels;
-#else
 	return GetCodecContext()->channels;
 #else
 	return GetCodecContext()->ch_layout.nb_channels;
@@ -101,7 +101,7 @@ bool AudioEncoder::AVCodecIsSupported(const QString& codec_name) {
 	return false;
 }
 
-void AudioEncoder::PrepareStream(AVStream* stream, AVCodecContext* codec_context, AVCodec* codec, AVDictionary** options, const std::vector<std::pair<QString, QString> >& codec_options,
+void AudioEncoder::PrepareStream(AVStream* stream, AVCodecContext* codec_context, const AVCodec* codec, AVDictionary** options, const std::vector<std::pair<QString, QString> >& codec_options,
 								 unsigned int bit_rate, unsigned int channels, unsigned int sample_rate) {
 
 	if(channels == 0) {
@@ -115,9 +115,6 @@ void AudioEncoder::PrepareStream(AVStream* stream, AVCodecContext* codec_context
 
 	codec_context->bit_rate = bit_rate;
 #if LIBAVCODEC_VERSION_MAJOR < 61
-#if SSR_USE_AV_CHANNEL_LAYOUT
-	av_channel_layout_default(&codec_context->ch_layout, channels);
-#else
 	codec_context->channels = channels;
 	codec_context->channel_layout = (channels == 1)? AV_CH_LAYOUT_MONO : AV_CH_LAYOUT_STEREO;
 #else
@@ -176,13 +173,11 @@ bool AudioEncoder::EncodeFrame(AVFrameWrapper* frame) {
 #  if LIBAVCODEC_VERSION_MAJOR < 61
 		assert(frame->GetFrame()->channels == GetChannels());
 #  else
-
-#if SSR_USE_AV_CHANNEL_LAYOUT
-	assert(frame->GetFrame()->ch_layout.nb_channels == GetCodecContext()->ch_layout.nb_channels);
-#else
-	assert(frame->GetFrame()->channels == GetCodecContext()->channels);
-#endif
-		assert(frame->GetFrame()->ch_layout.nb_channels == GetChannels());
+#    if SSR_USE_AV_CHANNEL_LAYOUT
+		assert(frame->GetFrame()->ch_layout.nb_channels == GetCodecContext()->ch_layout.nb_channels);
+#    else
+		assert(frame->GetFrame()->channels == GetCodecContext()->channels);
+#    endif
 #  endif /* LIBAVCODEC_VERSION_MAJOR < 61 */
 #endif
 #if SSR_USE_AVFRAME_SAMPLE_RATE
diff --git a/src/AV/Output/BaseEncoder.cpp b/src/AV/Output/BaseEncoder.cpp
index d2ef34c..e51876a 100644
--- a/src/AV/Output/BaseEncoder.cpp
+++ b/src/AV/Output/BaseEncoder.cpp
@@ -42,7 +42,7 @@ double ParseCodecOptionDouble(const QString& key, const QString& value, double m
 	return clamp(value_double, min, max) * multiply;
 }
 
-BaseEncoder::BaseEncoder(Muxer* muxer, AVStream* stream, AVCodecContext* codec_context, AVCodec* codec, AVDictionary** options) {
+BaseEncoder::BaseEncoder(Muxer* muxer, AVStream* stream, AVCodecContext* codec_context, const AVCodec* codec, AVDictionary** options) {
 
 	m_muxer = muxer;
 	m_stream = stream;
@@ -157,7 +157,7 @@ void BaseEncoder::IncrementPacketCounter() {
 	++lock->m_total_packets;
 }
 
-void BaseEncoder::Init(AVCodec* codec, AVDictionary** options) {
+void BaseEncoder::Init(const AVCodec* codec, AVDictionary** options) {
 
 	// open codec
 	if(avcodec_open2(m_codec_context, codec, options) < 0) {
